### PR TITLE
[SPARK-8266][SQL]add function translate

### PR DESCRIPTION
**Pull request by zhichao-li**
translate

### CODE REVIEW

**Comment by zhichao-li** [comment]
cc @chenghao-intel @adrian-wang

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala [hunk 1]
```diff
@@ -331,6 +331,16 @@ class StringExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     // scalastyle:on
   }
 
+  test("translate") {
+    checkEvaluation(StringTranslate("translate", "rnlt", "123"), "1a2s3ae")
```
Comment (adrian-wang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala]: how does this work? I think we should use literal.

Comment (zhichao-li) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala]: I guess it's due to there's an implicit ExpressionConversions for this as a shortcut.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala [hunk 2]
```diff
@@ -284,6 +284,66 @@ case class EndsWith(left: Expression, right: Expression)
...
+
+  override def inputTypes: Seq[DataType] = Seq(StringType, StringType, StringType)
+
+  override def children: Seq[Expression] = srcExpr :: matchingExpr :: replaceExpr :: Nil
+
+  override def eval(input: InternalRow): Any = {
+    val srcEval = srcExpr.eval(input)
+    if (srcEval != null) {
+      val matchingEval = matchingExpr.eval(input)
+      if (matchingEval != null) {
```
Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: you can use nullSafeEval

Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: oh, just found this don't have jt ..

**Comment by zhichao-li** [comment]
retest this please.

**Comment by zhichao-li** [comment]
retest this please.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 3]
```diff
@@ -507,6 +507,48 @@ public static UTF8String concatWs(UTF8String separator, UTF8String... inputs) {
     }
     return res;
   }
+  
+  private class Code {
```
Comment (chenghao-intel) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Nit: static class

**Comment by chenghao-intel** [comment]
Probably we need to make the cache for the `dict`, as in most of cases the `matching` and the `replace`  string are literal.

Anyway, I can leave it for further improvement.

@rxin @davies can we merge that first? As the code freeze in next 1 or 2 days.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala (hunk 1 above)
Comment (rxin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala]: It'd be better to not rely on that here

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 4]
```diff
@@ -507,6 +507,48 @@ public static UTF8String concatWs(UTF8String separator, UTF8String... inputs) {
...
+  public UTF8String translate(UTF8String matchingString, UTF8String replaceString) {
+    String srcStr = this.toString();
+    String matching = matchingString.toString();
+    String replace = replaceString.toString();
+    Map<Character, Code> dict = new HashMap<>();
+    int i = 0;
+    int j = 0;
+    while (i < matching.length() && j < replace.length()) {
+      // For multiple mapping, the first one wins.
+      if (null == dict.get(matching.charAt(i))) {
```
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: nit: In order to support Unicode characters that can't fit in a Char (using two Char in UTF-16), we should use code point  here. We could add a TODO for it now.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 5]
```diff
@@ -507,6 +507,48 @@ public static UTF8String concatWs(UTF8String separator, UTF8String... inputs) {
...
+    public final Character value;
+    public Code(Character value) {
+      this.value = value;
+    }
+    public boolean isEmpty(){
+      return value == null;
+    }
+  }
+
+  public UTF8String translate(UTF8String matchingString, UTF8String replaceString) {
```
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: It's better to passing in a HashMap here.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 6]
```diff
@@ -507,6 +507,48 @@ public static UTF8String concatWs(UTF8String separator, UTF8String... inputs) {
...
+    String srcStr = this.toString();
+    String matching = matchingString.toString();
+    String replace = replaceString.toString();
+    Map<Character, Code> dict = new HashMap<>();
```
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: We could use `0` as special Char, then `Code` is not needed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala (hunk 2 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: will be introduced in #7816

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 6 above)
Comment (zhichao-li) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Had that idea before but just in case we would pass `0` within the string. If that's acceptable I would update the code shortly.

Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: I think it's fine in practice.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala (hunk 2 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: It's merged, could you update to use TernaryExpression ?

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 7]
```diff
@@ -1718,6 +1718,18 @@ object functions {
...
   /**
+   * Translate any character in the srcString by a character in replaceString.
+   * The characters in replaceString is corresponding to the characters in matchingString.
+   * The translate will happen when any character in the string matching with the character
+   * in the matchingString.
+   *
+   * @group string_funcs
+   * @since 1.5.0
+   */
+  def translate(srcString: Column, matchingString: Column, replaceString: Column): Column =
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Should we use String for matching and replace, instead of Column?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala [hunk 8]
```diff
@@ -283,6 +283,118 @@ case class EndsWith(left: Expression, right: Expression)
...
+
+  def buildDict(matchingString: UTF8String, replaceString: UTF8String)
+    : JMap[Character, Character] = {
+    val matching = matchingString.toString()
+    val replace = replaceString.toString()
+    val dict = new HashMap[Character, Character]()
+    var i = 0
+    var j = 0
+    while (i < matching.length() && j < replace.length()) {
+      // For multiple mapping, the first one wins.
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: Is this the designed behavior or come from Hive?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala [hunk 9]
```diff
@@ -283,6 +283,118 @@ case class EndsWith(left: Expression, right: Expression)
...
+      if (null == dict.get(matching.charAt(i))) {
+        dict.put(matching.charAt(i), replace.charAt(j))
+      }
+      i += 1
+      j += 1
+    }
+    while (i < matching.length()) {
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: I think one loop should be fine:
```
while (i < matching.length()) {
  val rep = if (i < replace.length()) replace.charAt(i) else '0'
    if (null == dict.get(matching.charAt(i))) {
        dict.put(matching.charAt(i), rep)
     }
}
```

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 10]
```diff
@@ -508,6 +508,43 @@ public static UTF8String concatWs(UTF8String separator, UTF8String... inputs) {
     return res;
   }
 
+  private Map buildTranslateMap(UTF8String matchingString, UTF8String replaceString) {
```
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Why we have two function for buildTranslateMap?  another one is `buildDict`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala (hunk 8 above)
Comment (zhichao-li) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: This is come from hive and there's compatibilitysuite to guard against this.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 10 above)
Comment (zhichao-li) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: my bad, would delete this one.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 7 above)
Comment (zhichao-li) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Either is ok but seems like the others opt for String. would update this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala (hunk 9 above)
Comment (zhichao-li) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: Agree.  We only respect matching in this case.

**Comment by davies** [comment]
build/sbt-launch-0.13.7.jar.part should be removed

**Comment by zhichao-li** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala [hunk 11]
```diff
@@ -350,6 +351,78 @@ case class EndsWith(left: Expression, right: Expression)
...
+
+    ctx.addMutableState(classNameUTF8String, termLastMatching, s"${termLastMatching} = null;")
+    ctx.addMutableState(classNameUTF8String, termLastReplace, s"${termLastReplace} = null;")
+    ctx.addMutableState(classNameDict, termDict, s"${termDict} = null;")
+
+    nullSafeCodeGen(ctx, ev, (src, matching, replace) => {
+      s"""if (!${matching}.equals(${termLastMatching}) ||
+        !${replace}.equals(${termLastReplace})) {
+        // matching or replace value changed
+        ${termLastMatching} = ${matching};
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: `${matching}.clone()`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala [hunk 12]
```diff
@@ -350,6 +351,78 @@ case class EndsWith(left: Expression, right: Expression)
...
+      dict = StringTranslate.buildDict(lastMatching, lastReplace)
+    }
+    srcEval.asInstanceOf[UTF8String].translate(dict)
+  }
+
+  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
+    val termLastMatching = ctx.freshName("lastMatching")
+    val termLastReplace = ctx.freshName("lastReplace")
+    val termDict = ctx.freshName("dict")
+    val classNameUTF8String = classOf[UTF8String].getCanonicalName
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: UTF8String is already imported, so just use `UTF8String` is better.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala [hunk 13]
```diff
@@ -350,6 +351,78 @@ case class EndsWith(left: Expression, right: Expression)
...
+    val classNameDict = classOf[JMap[Character, Character]].getCanonicalName
+
+    ctx.addMutableState(classNameUTF8String, termLastMatching, s"${termLastMatching} = null;")
+    ctx.addMutableState(classNameUTF8String, termLastReplace, s"${termLastReplace} = null;")
+    ctx.addMutableState(classNameDict, termDict, s"${termDict} = null;")
+
+    nullSafeCodeGen(ctx, ev, (src, matching, replace) => {
+      s"""if (!${matching}.equals(${termLastMatching}) ||
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: This will always be true, if matchingExpr and replaceExpr are literals (this is the common use case, I think), we can avoid this check by having a special case for both matchingExpr and replaceExpr are foldable.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 14]
```diff
@@ -1927,6 +1927,17 @@ object functions {
...
 
+  /* Translate any character in the srcString by a character in replaceString.
+  * The characters in replaceString is corresponding to the characters in matchingString.
+  * The translate will happen when any character in the string matching with the character
+  * in the matchingString.
+  *
+  * @group string_funcs
+  * @since 1.5.0
+  */
+  def translate(srcString: Column, matchingString: String, replaceString: String): Column =
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Using `src` instead `strString` will be better.

File: sql/core/src/test/scala/org/apache/spark/sql/StringFunctionsSuite.scala [hunk 15]
```diff
@@ -128,6 +128,18 @@ class StringFunctionsSuite extends QueryTest {
...
 
+  test("string translate") {
+    val df = Seq(("translate", "")).toDF("a", "b")
+
+    checkAnswer(df.select(translate($"a", "rnlt", "123")), Row("1a2s3ae"))
+    checkAnswer(df.select(translate($"a", "", "123")), Row("translate"))
+    checkAnswer(df.select(translate($"a", "rnlt", "")), Row("asae"))
+    // scalastyle:off
+    // non ascii characters are not allowed in the source code, so we disable the scalastyle.
+    checkAnswer(df.select(translate(lit("花花世界"), "花界", "ab")), Row("aa世b"))
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/StringFunctionsSuite.scala]: This is an end-to-end test for DataFrame or SQL parse, these special cases are not necessary here. Instead, we should have one fro SQL parser (using selectExpr())

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 16]
```diff
@@ -741,6 +742,21 @@ public static UTF8String concatWs(UTF8String separator, UTF8String... inputs) {
     return res;
   }
 
+  // TODO: Need to use `Code Point` here instead of Char in case the character longer than 2 bytes
```
Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: We could have a fast version (using byte), and a slow version (using int as code point), that could be a separate PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala [hunk 17]
```diff
@@ -350,6 +351,78 @@ case class EndsWith(left: Expression, right: Expression)
...
+object StringTranslate {
+
+  def buildDict(matchingString: UTF8String, replaceString: UTF8String)
+    : JMap[Character, Character] = {
+    val matching = matchingString.toString()
+    val replace = replaceString.toString()
+    val dict = new HashMap[Character, Character]()
+    var i = 0
+    while (i < matching.length()) {
+      val rep = if (i < replace.length()) replace.charAt(i) else '0'
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: Using `0`, not `'0'`

File: unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java [hunk 18]
```diff
@@ -392,6 +394,26 @@ public void levenshteinDistance() {
...
+      )));
+    assertEquals(fromString("translate"),
+      fromString("translate").translate(new HashMap<Character, Character>()));
+    assertEquals(fromString("asae"),
+      fromString("translate").translate(ImmutableMap.of(
+        'r', '0',
+        'n', '0',
+        'l', '0',
+        't', '0'
+      )));
```
Comment (davies) [inline unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java]: We should put all the corner cases here, for example, Chinese, empty dict, or empty string.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala (hunk 11 above)
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: You mean the underlying object that `UTF8String.base` pointed will be changed in the next iteration? So we need a deep copy here?

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 16 above)
Comment (zhichao-li) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Not sure if I can get you. 
For the byte version you mean passing in a Map of Array[Byte] -> Array[Byte]? 
4 byte for int, would it be acceptable for a code point of UTF8?

Comment (davies) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: If all the characters are less than 127, then we could use Map[Byte][Byte]( even Array[byte][128]), otherwise use Map[Int][Int]

**Comment by zhichao-li** [comment]
retest this please.

File: unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java [hunk 19]
```diff
@@ -148,26 +150,26 @@ public void concatWsTest() {
       EMPTY_UTF8,
       concatWs(sep, EMPTY_UTF8));
     assertEquals(
-      fromString("ab"),
-      concatWs(sep, fromString("ab")));
+            fromString("ab"),
```
Comment (davies) [inline unsafe/src/test/java/org/apache/spark/unsafe/types/UTF8StringSuite.java]: Why this? Could you update the code style of Java in your IDE? using indent as 2 space.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala [hunk 20]
```diff
@@ -349,6 +351,78 @@ case class EndsWith(left: Expression, right: Expression)
...
+case class StringTranslate(srcExpr: Expression, matchingExpr: Expression, replaceExpr: Expression)
+  extends TernaryExpression with ImplicitCastInputTypes {
+
+  @transient private var lastMatching: UTF8String = _
+  @transient private var lastReplace: UTF8String = _
+  @transient private var dict: JMap[Character, Character] = _
+
+  override def nullSafeEval(srcEval: Any, matchingEval: Any, replaceEval: Any): Any = {
+    if (matchingEval != lastMatching || replaceEval != lastReplace) {
+      lastMatching = matchingEval.asInstanceOf[UTF8String]
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: .clone()

Comment (zhichao-li) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: @davies just not quite sure why we need clone() here, seems like the UTF8String implementation is kind of like String which do the "copy on write". I guess we can somehow treat it as a constant value?

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: We recently changed to do not copy the bytes when call UnsafeRow.getUTF8String(), so whenever we want to hold a UTF8String, we need to do a copy.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala [hunk 21]
```diff
@@ -349,6 +351,78 @@ case class EndsWith(left: Expression, right: Expression)
...
+    val termLastReplace = ctx.freshName("lastReplace")
+    val termDict = ctx.freshName("dict")
+    val classNameDict = classOf[JMap[Character, Character]].getCanonicalName
+
+    ctx.addMutableState("UTF8String", termLastMatching, s"${termLastMatching} = null;")
+    ctx.addMutableState("UTF8String", termLastReplace, s"${termLastReplace} = null;")
+    ctx.addMutableState(classNameDict, termDict, s"${termDict} = null;")
+
+    nullSafeCodeGen(ctx, ev, (src, matching, replace) => {
+      s"""if (!(${matchingExpr.foldable} && ${replaceExpr.foldable} && ${termDict} != null) ||
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala]: This is confusing, when matchingExpr.foldable is true and is true, it will still check !${matching}.equals(${termLastMatching})

```
val check = if (matchingExpr.foldable && replaceExpr.foldable) {
 s"${termDict} == null"
} else {
  s"!${matching}.equals(${termLastMatching}) || !${replace}.equals(${termLastReplace})"
}

s"""
if ($check) {
 ...
}
"""
```

**Comment by davies** [comment]
@zhichao-li Thanks for working on this, it's very closing, will merge it once you address the last comment.

**Comment by davies** [comment]
LGTM, waiting for tests.

**Comment by zhichao-li** [comment]
retest this please. It fail in not related test.

**Comment by davies** [comment]
Merged into master and 1.5 branch, thanks!
